jxl group

damdam307 opened the pull request.

veluca93 commented on README.md:
Please don't delete this file.

veluca93 commented on multilayer_jxl/out/CONTRIBUTING.md:
And don't rename this file.

veluca93 commented on multilayer_jxl/out/README.md:
```diff
@@ -0,0 +1,37 @@
+# Computer Science Olympiads Student Projects
```
Leave the original readme out

veluca93 commented on multilayer_jxl/out/build.sh:
```diff
@@ -0,0 +1,59 @@
+
```
No newline here

veluca93 commented:
Please re-write the commit history to remove fixup commits

veluca93 commented on Multilayer_JXL/README.md:
````diff
@@ -0,0 +1,26 @@
+The ```main.py``` file acts as an API for the two steps of compressing the image: training and JXL encoding.
+To successfully run the script, use the command ```python main.py [arguments]```.
+The following arguments are supported:
+- `<filename>`: necessary argument that specifies the input file. if the argument is written as `<folder_path>/` the script will try to compress all the files in the folder. When an error occurs during compression the scripts writes the error and continues to the next file. Supported types are [all Pyllow supported file types](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.ht ...
````
Pillow, not Pyllow

veluca93 commented on Multilayer_JXL/README.md:
````diff
@@ -0,0 +1,26 @@
+To successfully run the script, use the command ```python main.py [arguments]```.
+The following arguments are supported:
+- `<filename>`: necessary argument that specifies the input file. if the argument is written as `<folder_path>/` the script will try to compress all the files in the folder. When an error occurs during compression the scripts writes the error and continues to the next file. Supported types are [all Pyllow supported file types](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.ht ...
+- `<layer>`: must go after (not necessarily directly after) the input parameter. is an integer 0-4 that describes how many layers will be ancoded in the JXL file.
````
`<number_of_layers>`, must go after the input parameter. Represents how many layers will be incoded in the JXL file. Valid range 1 to 4.

(I imagine 0 doesn't make a lot of sense?)

veluca93 commented on Multilayer_JXL/README.md:
```diff
@@ -0,0 +1,26 @@
+The following arguments are supported:
+- `<filename>`: necessary argument that specifies the input file. if the argument is written as `<folder_path>/` the script will try to compress all the files in the folder. When an error occurs during compression the scripts writes the error and continues to the next file. Supported types are [all Pyllow supported file types](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.ht ...
+- `<layer>`: must go after (not necessarily directly after) the input parameter. is an integer 0-4 that describes how many layers will be ancoded in the JXL file.
+- `-o <filename>`: specifies an output path for the JXL file. If the input is a folder, output must also be a folder (`<folder_path>/`) or it will be overridden with `out/`. If the input is a single file, output can be a filename or a folder, in the second case the final path will be `<folder_path>/<filename>.jxl`.
```
"or it will be overridden with" -> "defaults to"

veluca93 commented on Multilayer_JXL/README.md:
```diff
@@ -0,0 +1,26 @@
+- `<filename>`: necessary argument that specifies the input file. if the argument is written as `<folder_path>/` the script will try to compress all the files in the folder. When an error occurs during compression the scripts writes the error and continues to the next file. Supported types are [all Pyllow supported file types](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.ht ...
+- `<layer>`: must go after (not necessarily directly after) the input parameter. is an integer 0-4 that describes how many layers will be ancoded in the JXL file.
+- `-o <filename>`: specifies an output path for the JXL file. If the input is a folder, output must also be a folder (`<folder_path>/`) or it will be overridden with `out/`. If the input is a single file, output can be a filename or a folder, in the second case the final path will be `<folder_path>/<filename>.jxl`.
+- `-l <lambda>` specifies the float value of the lambda parameter. Defaults to 0
```
Describing what those do would probably help.

veluca93 replied:
(also for the parameters below)

veluca93 commented on Multilayer_JXL/README.md:
```diff
@@ -0,0 +1,26 @@
+- `-ws <turns>` specifies how many rounds of wasserstein distortion training will be applied. Defaults to 10000
+- `-p <patience>` specifies the integer value of the patience parameter. Defaults to 100000
+- `-d <delta>` specifies the float value of the delta parameter. Defaults to 1e-6
+- `-prefix <prefix>` specifies the folder to use for the output files. Defaults to `out/`. It is useless to set this if you don't care about the intermediate files.
```
"specifies the folder to store intermediate files in"

veluca93 commented on Multilayer_JXL/README.md:
```diff
@@ -0,0 +1,26 @@
+- `-p <patience>` specifies the integer value of the patience parameter. Defaults to 100000
+- `-d <delta>` specifies the float value of the delta parameter. Defaults to 1e-6
+- `-prefix <prefix>` specifies the folder to use for the output files. Defaults to `out/`. It is useless to set this if you don't care about the intermediate files.
+- `-space <space>` specifies the training space to use for the JXL encoding. Defaults to "rgb", can be also "xyb" or "dct"
```
training space -> variables?

veluca93 commented on README.md:
```diff
@@ -7,4 +7,4 @@ in the [Swiss Olympiads in Informatics](https://soi.ch).
 This is not an officially supported Google product. This project is not
 eligible for the [Google Open Source Software Vulnerability Rewards
-Program](https://bughunters.google.com/open-source-security).
+Program](https://bughunters.google.com/open-source-security).x
```
Please don't modify this file.

veluca93 commented on Multilayer_JXL/README.md:
```diff
@@ -0,0 +1,26 @@
+- `-d <delta>` specifies the float value of the delta parameter. Defaults to 1e-6
+- `-prefix <prefix>` specifies the folder to use for the output files. Defaults to `out/`. It is useless to set this if you don't care about the intermediate files.
+- `-space <space>` specifies the training space to use for the JXL encoding. Defaults to "rgb", can be also "xyb" or "dct"
+- `-c` enables the use of the settings from the settings.conf file. Each line in the config file should be in the format `<layer>: <value>`, where layer is an integer from 0 to 4 which refers to the 2^<layer>x downsampled layer and <value> is the corresponding setting.
```
I don't think this is particularly clear.

veluca93 commented on Multilayer_JXL/build.sh:
```diff
@@ -0,0 +1,59 @@
+
```
Extra newline

veluca93 commented on Multilayer_JXL/faster_wasserstein_vgg16.py:
```diff
@@ -0,0 +1,96 @@
+    Parameters
+    ----------
+    image
+        image to be converted in [H, W, C] format
```
"to be converted"?

veluca93 commented on Multilayer_JXL/image_converter_utils.py:
```diff
@@ -0,0 +1,346 @@
+    return srgb01
+
+def add_y_to_b(img):
+    x = img[:, :, 0, :, :]
```
note: if those lines were `img[..., 0, :, :]`, this would support more shapes.

veluca93 commented on Multilayer_JXL/jxl_layered_encoder.cc:
```diff
@@ -0,0 +1,193 @@
+ */
+vector<float> loadImage(const string& path,
+                        size_t num_channels,
+                        size_t bytes_per_pixel,
```
I don't believe you need this parameter (or should multiply by it) since you return a vector of floats.

damdam307 replied:
We noticed the parameter was useless, but for some reason just removing it made the encoding fail. I'll look into why.

veluca93 commented on Multilayer_JXL/jxl_layered_encoder.cc:
```diff
@@ -0,0 +1,193 @@
+
+            // Set resampling factor (power of 2 downscaling)
+            int scale_factor = 1 << (N - i);
+            cout << "Adding frame " << i << " with scale factor " << scale_factor << endl;
```
I'm _fairly_ sure that the patches to libjxl will cause the resulting jxl image to only add the last two images together. To fix that, one probably needs to modify the line `if (!cparams.already_downsampled) {` to instead check whether this is not the first frame, or something similar.

veluca93 commented on Multilayer_JXL/jxl_layered_encoder.cc:
```diff
@@ -0,0 +1,193 @@
+
+        // Clean up encoder
+        JxlEncoderDestroy(enc);
+    } catch (const exception& e) {
```
I don't think anything can actually return an exception in this try..catch block, can you get rid of it?

veluca93 commented on Multilayer_JXL/jxl_layered_encoder.cc:
```diff
@@ -0,0 +1,193 @@
+
+        // Save compressed data to file
+        size_t compressed_size = next_out - compressed.data();
+        std::ofstream out(output_path, std::ios::binary);
```
Please check if opening `out` failed.

veluca93 commented on Multilayer_JXL/README.md:
```diff
@@ -0,0 +1,26 @@
+- `-l2 <turns>` specifies how many rounds of l2 training will be applied. Defaults to 90000
+- `-ws <turns>` specifies how many rounds of wasserstein distortion training will be applied. Defaults to 10000
+- `-p <patience>` specifies the integer value of the patience parameter. Defaults to 100000
+- `-d <delta>` specifies the float value of the delta parameter. Defaults to 1e-6
```
This doesn't seem to have a flag.

veluca93 commented on Multilayer_JXL/README.md:
````diff
@@ -0,0 +1,26 @@
+# How to run
+The ```main.py``` file acts as an API for the two steps of compressing the image: training and JXL encoding.
+To successfully run the script, use the command ```python main.py [arguments]```.
+The following arguments are supported:
````
I think it might be better to instruct people to run `python main.py --help` instead, and making sure that the documentation of the flags is clear enough there.

veluca93 commented on Multilayer_JXL/losses.py:
```diff
@@ -0,0 +1,188 @@
+        a weighting of shape [C] for the different channels
+    """
+    img_batched = jnp.expand_dims(img, 0)
+    img_convoluted = lax.conv_general_dilated(
```
nit: "convolved", not "convoluted"

veluca93 commented on Multilayer_JXL/main.py:
```diff
@@ -0,0 +1,171 @@
+# Parse arguments
+args = parse_arguments()
+
+# Extract values from args
```
I'd prefer not to do this.

veluca93 commented on Multilayer_JXL/main.py:
```diff
@@ -0,0 +1,171 @@
+    
+    return parser.parse_args()
+
+# Parse arguments
```
can we move all of this code in a `main()` function, and call that under `if __name__ == "__main__"`?

veluca93 commented on Multilayer_JXL/main.py:
```diff
@@ -0,0 +1,171 @@
+    try:
+        with Image.open(f, 'r') as file:
+            target = jnp.asarray(file, dtype=jnp.float32) / 255.
+            target = jnp.array(target[:(target.shape[0]//(2**(2+layers)))*(2**(2+layers)), :(target.shape[1]//(2**(2+layers)))*(2**(2+layers)), :])
```
This line would benefit from creating a variable with `2**(2+layers)`.

veluca93 commented on Multilayer_JXL/train.py:
```diff
@@ -0,0 +1,212 @@
+from optimizer_values import RGBOptimizerValues, XYBOptimizerValues, XYBDCTOptimizerValues
+from codex.loss import load_vgg16_model
+
+# Cache jit compilation
```
At least some of those variables are already set in `main.py`.

42kangaroo replied:
deleted them from main

veluca93 commented on Multilayer_JXL/train.py:
```diff
@@ -0,0 +1,212 @@
+        raise Exception(f"Unknown base type: {base}")
+    log2_sigma = jnp.full((H, W), log2_sigma_value, dtype=jnp.float32)
+
+    loss, (ws_loss, compression_losses) = loss_fn(
```
I am not sure printing loss values at initialization is extremely useful, except for debugging.

veluca93 commented on Multilayer_JXL/train.py:
```diff
@@ -0,0 +1,212 @@
+                use_l2,
+            )
+            print(
+                f"Step {step + 1}, loss {loss}, {'l2 loss' if step < l2_loops else 'wloss'} {ws_loss}, closs {sum(compression_losses)} ({jnp.array(compression_losses).tolist()})"
```
It might be a good idea to format this line better, and possibly to set things up so that it's overwritten except every 2000 or so epochs, i.e.:
```
            closses = ", ".join(f'{x:5.2f}' for x in np.array(compression_losses).tolist())
            print(
                f"Step {step+1:9d}, loss {loss:5.2f}, wloss {ws_loss:5.2f}, closs {sum(compression_losses):5.2f} ({closses}) lambda {float(llambda):5.2f}", end='\n' if step % 2000 == 0 else '\r'
            )
```

42kangaroo commented:
applied every review for multilayer model and some minor ones for main and documentation @damdam307 and @NoemiGambirasio you can look at libjxl custom encoder and some more in main.

veluca93 commented:
> applied every review for multilayer model and some minor ones for main and documentation @damdam307 and @NoemiGambirasio you can look at libjxl custom encoder and some more in main.

Please rebase this PR on top of `main`, or github will get confused.

veluca93 commented on Multilayer_JXL/main.py:
```diff
@@ -0,0 +1,175 @@
+import os
+# Suppress XLA/CUDA warning messages
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'  # Suppress TensorFlow/XLA logs
+os.environ['XLA_FLAGS'] = '--xla_gpu_cuda_data_dir=/usr/local/cuda'
```
This doesn't quite look right, are you sure it is helpful/correct?

damdam307 commented:
Fixed all issues related to main and jxl custom encoder but [this one on fixing the patch](https://github.com/paradigms-of-intelligence/cs-olympiads-student-projects/pull/2#discussion_r2296128650) and [this one on transferring flag documentation to --help](https://github.com/paradigms-of-intelligence/cs-olympiads-student-projects/pull/2#discussion_r2296132108).

veluca93 commented on Multilayer_JXL/jxl_layered_encoder.cc:
```diff
@@ -0,0 +1,191 @@
+
+        // Add frame data to encoder
+        JxlEncoderAddImageFrame(frame_settings, &pixel_format,
+                                frames[i].data(), widths[i]*heights[i]*num_channels*bytes_per_pixel);
```
Looks like this file needs to be autoformatted.

damdam307 commented:
I removed the jax silencers, formatted a bit jxl_layered_encoder.cc, fixed a small bug with intermediate file naming and added some info in --help. If the --help is good I'll remove the flags from the readme and instruct to instead run the command.
The jxl encoder is still broken for 3+ levels, it is indeed needed to change the patch, though there are a few bugs in the cpp file.

veluca93 commented:
I'd say to make the cpp file just return an error for a number of levels that is not 2. Also, please use clang-format to format it.

The documentation of the flags looks good.

damdam307 commented:
added the 3+ layer error, formatted with clang-format and changed readme to instruct to run --help

veluca93 commented:
Perfect. One last thing: currently the commits are all authored by you. Can you add appropriate Co-Authored-By lines for the other people?

damdam307 commented:
Done. I added you too since you worked on the libjxl patch.

damdam307 committed:
Add Multilayer models

Co-authored-by: NoemiGambirasio <noemi.gambirasio@gmail.com>
Co-authored-by: Yaël Arn <yael.arn@bluewin.ch>
Co-authored-by: Luca Versari <veluca@google.com>

damdam307 committed:
Add libjxl custom encoder

Co-authored-by: NoemiGambirasio <noemi.gambirasio@gmail.com>
Co-authored-by: Yaël Arn <yael.arn@bluewin.ch>
Co-authored-by: Luca Versari <veluca@google.com>

damdam307 committed:
Add main, documentation and build script

Co-authored-by: NoemiGambirasio <noemi.gambirasio@gmail.com>
Co-authored-by: Yaël Arn <yael.arn@bluewin.ch>
Co-authored-by: Luca Versari <veluca@google.com>

damdam307 commented:
Removed Jason from the co-authors since he hadn't signed the CLA.
